File: DailyChallenges/Scala/LeetCode_739.scala
object Solution {
    def dailyTemperatures(temperatures: Array[Int]): Array[Int] = {
        val ans = Array.ofDim[Int](temperatures.length)
        var l = 0 :: Nil;
        for (i <- 1 until temperatures.length){
            while(l.nonEmpty && temperatures(l.head) < temperatures(i)){
                ans(l.head) = i - l.head
                l = l.tail
            }
            l = i :: l
        }
        ans
    }
}

File: DailyChallenges/CPP/LeetCode_739.cpp
// 739. Daily Temperatures @ 2022/12/18
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        // max stack
        vector<int> ans(temperatures.size(), 0);
        vector<int> stack; stack.push_back(0);
        for (int i = 1; i<temperatures.size(); ++i){
            while (!stack.empty() && temperatures[stack.back()] < temperatures[i]){
                ans[stack.back()] = i - stack.back();
                stack.pop_back();
            }
            stack.push_back(i);
        }
        return ans;
    }
};
